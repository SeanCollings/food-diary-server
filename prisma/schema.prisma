// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

// https://www.prisma.io/nestjs
// https://docs.nestjs.com/recipes/prisma
// https://blog.logrocket.com/how-to-use-nestjs-prisma/

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model User {
  id        Int        @id @default(autoincrement())
  name      String
  email     String     @unique
  password  String
  avatar    String?
  createdAt DateTime   @default(now())
  lastLogin DateTime   @default(now())
  shareLink ShareLink?
  diaryDay  DiaryDay[]

  userPreferenceShowDayStreak       Boolean  @default(true)
  userPreferenceShowWeeklyExcercise Boolean?
  userPreferenceShowWeeklyWater     Boolean?

  statLastActivity DateTime?
  statDayStreak    Int?
}

model ShareLink {
  id       Int     @id @default(autoincrement())
  user     User    @relation(fields: [userId], references: [id])
  userId   Int     @unique
  link     String  @unique
  isShared Boolean
}

model DiaryDay {
  id                       String  @id @default(uuid())
  user                     User    @relation(fields: [userId], references: [id])
  userId                   Int
  date                     String
  mealBreakfast            Json?
  mealSnack1               Json?
  mealLunch                Json?
  mealSnack2               Json?
  mealDinner               Json?
  wellnessWater            Float?
  wellnessTeaCoffee        Float?
  wellnessAlcohol          Float?
  wellnessExcercise        String?
  wellnessExcerciseDetails String?

  hasMealBreakfast     Boolean?
  hasMealSnack1        Boolean?
  hasMealLunch         Boolean?
  hasMealSnack2        Boolean?
  hasMealDinner        Boolean?
  hasWellnessExcercise Boolean?

  @@unique([userId, date])
}
