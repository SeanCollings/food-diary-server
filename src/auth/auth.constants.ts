export const jwtContants = {
  secret: process.env.JWT_SECRET,
};

export const SIGN_OPTION_EXPIRES_IN = '30d';

export const GOOGLE_RECAPTCHA_PASS_SCORE = 0.5;

export const SCRYPT_KEYLEN = 32;
