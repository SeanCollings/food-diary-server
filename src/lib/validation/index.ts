import { trim } from '@/utils/string-utils';
import {
  EMAIL_MIN_LENGTH,
  EMAIL_REGEX,
  INPUT_MAX_LENGTH,
  PASSWORD_MIN_LENGTH,
} from '@/lib/constants/validation/validation.constants';

export const validateEmail = (email: string) => {
  const trimEmail = trim(email);

  if (!trimEmail?.length) {
    return false;
  } else if (!EMAIL_REGEX.test(trimEmail)) {
    return false;
  }

  return (
    trimEmail.length >= EMAIL_MIN_LENGTH && trimEmail.length <= INPUT_MAX_LENGTH
  );
};

export const validatePassword = (password: string) => {
  const trimPassword = trim(password);

  if (!trimPassword?.length) {
    return false;
  }

  return (
    trimPassword.length >= PASSWORD_MIN_LENGTH &&
    trimPassword.length <= INPUT_MAX_LENGTH
  );
};

export const validateLogin = ({
  email,
  password,
}: {
  email: string;
  password: string;
}) => validateEmail(email) && validatePassword(password);
