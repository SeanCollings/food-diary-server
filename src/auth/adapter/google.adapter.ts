import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { HttpService } from '@nestjs/axios';
import { catchError, firstValueFrom } from 'rxjs';
import { AxiosError } from 'axios';
import { RecaptchaResponse } from '../types';
import { GOOGLE_RECAPTCHA_PASS_SCORE } from '../auth.constants';

@Injectable()
export class GoogleAdapter {
  private recaptchaVerifyUrl: string;
  private recaptchaSecretKey: string;

  constructor(
    private readonly httpService: HttpService,
    private readonly configService: ConfigService,
  ) {
    this.recaptchaVerifyUrl =
      this.configService.get('RECAPTCHA_VERIFY_URL') || '';
    this.recaptchaSecretKey =
      this.configService.get('RECAPTCHA_SECRET_KEY') || '';
  }

  async verifySite(token: string) {
    const params = {
      secret: this.recaptchaSecretKey,
      response: token,
    };

    const { data } = await firstValueFrom(
      this.httpService
        .post<RecaptchaResponse>(this.recaptchaVerifyUrl, null, { params })
        .pipe(
          catchError((error: AxiosError) => {
            console.error('recaptcha error:', error.response?.data);
            throw new Error('A verification error ocurred!');
          }),
        ),
    );

    if (!data.success) {
      throw new Error(
        `Verification failed for ${data.action}: ${(
          data['error-codes'] ?? []
        ).join()}`,
      );
    }

    return data.score >= GOOGLE_RECAPTCHA_PASS_SCORE;
  }
}
