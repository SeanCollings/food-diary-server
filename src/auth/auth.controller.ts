import {
  Body,
  Controller,
  Post,
  Request,
  UseGuards,
  BadRequestException,
} from '@nestjs/common';
import { AuthService } from '@/auth/auth.service';
import { LocalAuthGuard } from '@/auth//local-auth.guard';
import { RequestWithUser } from './types';
import { CreateUserDTO, LoginUserDTO, ResetPasswordDto } from './dtos';

@Controller('auth')
export class AuthController {
  constructor(private authService: AuthService) {}

  @Post('/signup')
  async createUser(@Body() body: CreateUserDTO) {
    try {
      return await this.authService.signup(body);
    } catch (err) {
      console.error('[auth::signup]::', err.message);
      throw new BadRequestException(err.message || 'Something went wrong');
    }
  }

  @Post('/login')
  @UseGuards(LocalAuthGuard)
  async login(@Request() req: RequestWithUser, @Body() body: LoginUserDTO) {
    try {
      const { id, email } = req?.user || {};
      return await this.authService.login({ id, email, token: body.token });
    } catch (err) {
      console.error('[auth::login]::', err.message);
      throw new BadRequestException(err.message || 'Something went wrong');
    }
  }

  @Post('/reset')
  async resetPassword(@Body() body: ResetPasswordDto) {
    try {
      const { email, token } = body;
      return await this.authService.resetPassword({ email, token });
    } catch (err) {
      console.error('[auth::reset]::', err.message);
      throw new BadRequestException(err.message || 'Something went wrong');
    }
  }
}
