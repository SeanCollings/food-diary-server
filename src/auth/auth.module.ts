import { UsersModule } from '@/users/users.module';
import { Module } from '@nestjs/common';
import { AuthController } from '@/auth/auth.controller';
import { AuthService } from '@/auth/auth.service';
import { PassportModule } from '@nestjs/passport';
import { JwtModule } from '@nestjs/jwt';
import { jwtContants, SIGN_OPTION_EXPIRES_IN } from '@/auth/auth.constants';
import { LocalStrategy } from '@/auth/local.strategy';
import { JwtStrategy } from '@/auth/jwt.strategy';
import { UsersService } from '@/users/users.service';
import { PrismaService } from '@/prisma.service';

@Module({
  imports: [
    UsersModule,
    PassportModule,
    JwtModule.register({
      secret: jwtContants.secret,
      signOptions: { expiresIn: SIGN_OPTION_EXPIRES_IN },
    }),
  ],
  controllers: [AuthController],
  providers: [
    AuthService,
    LocalStrategy,
    JwtStrategy,
    UsersService,
    PrismaService,
  ],
  exports: [AuthService],
})
export class AuthModule {}
