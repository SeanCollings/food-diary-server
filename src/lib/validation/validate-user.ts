import { promisify } from 'util';
import { scrypt as _scrypt } from 'crypto';
import { User } from '@prisma/client';
import { SCRYPT_KEYLEN } from '@/auth/auth.constants';

const scrypt = promisify(_scrypt);

export const isValidPassword = async (user: User | null, password: string) => {
  if (!user) {
    return null;
  }

  const [salt, storedHash] = user.password.split('.');
  const hash = (await scrypt(password, salt, SCRYPT_KEYLEN)) as Buffer;

  if (storedHash === hash.toString('hex')) {
    return user;
  }

  return null;
};
