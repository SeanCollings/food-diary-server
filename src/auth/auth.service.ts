import { Injectable } from '@nestjs/common';
import { UsersService } from '@/users/users.service';
import { randomBytes, scrypt as _scrypt } from 'crypto';
import { promisify } from 'util';
import { JwtService } from '@nestjs/jwt';
import { PrismaService } from '@/prisma.service';
import { GoogleAdapter } from './adapter/google.adapter';
import { CreateUserDTO, ResetPasswordDto } from './dtos';

const scrypt = promisify(_scrypt);

@Injectable()
export class AuthService {
  constructor(
    private readonly userService: UsersService,
    private jwtService: JwtService,
    private prisma: PrismaService,
    private googleAdapter: GoogleAdapter,
  ) {}

  async validateUser(email: string, password: string) {
    const user = await this.userService.findOne(email);

    if (user) {
      const [salt, storedHash] = user.password.split('.');
      const hash = (await scrypt(password, salt, 32)) as Buffer;

      if (storedHash === hash.toString('hex')) {
        return user;
      }
    }

    return null;
  }

  async signup({ email, name, password, token }: CreateUserDTO) {
    const isVerified = await this.googleAdapter.verifySite(token);

    if (!isVerified) {
      throw new Error('Something went wrong!');
    }

    const user = await this.userService.findOne(email);

    if (user) {
      throw new Error('Email in use!');
    }

    const salt = randomBytes(8).toString('hex');
    const hash = (await scrypt(password, salt, 32)) as Buffer;
    const result = salt + '.' + hash.toString('hex');

    return this.prisma.user.create({
      data: { email, name, password: result },
    });
  }

  async login(args: { email: string; id: number; token: string }) {
    const isVerified = await this.googleAdapter.verifySite(args.token);

    if (!isVerified) {
      throw new Error('Something went wrong!');
    }

    await this.prisma.user.update({
      where: { id: args.id },
      data: { lastLogin: new Date() },
    });

    const payload = { email: args.email, sub: args.id };

    return {
      access_token: this.jwtService.sign(payload),
    };
  }

  async resetPassword(args: ResetPasswordDto) {
    const isVerified = await this.googleAdapter.verifySite(args.token);

    if (!isVerified) {
      throw new Error('Something went wrong!');
    }

    return { message: 'Reset link sent' };
  }
}
