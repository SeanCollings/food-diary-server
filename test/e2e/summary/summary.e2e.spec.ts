import { Test, TestingModule } from '@nestjs/testing';
import { INestApplication } from '@nestjs/common';
import * as request from 'supertest';
import { SummaryModule } from '@/summary/summary.module';
import { DeepMockProxy, mockDeep } from 'jest-mock-extended';
import { PrismaClient } from '@prisma/client';
import { PrismaService } from '@/prisma.service';
import { AppModule } from '@/app.module';
import { AuthModule } from '@/auth/auth.module';
import { UsersService } from '@/users/users.service';
import { isValidPassword } from '@/lib/validation/validate-user';
import { GoogleAdapter } from '@/auth/adapter/google.adapter';

import { mockLoginResponseBody } from '../e2e.setup';
import { user_fixture, diaryDays_fixture } from '../__fixtures__';

jest.mock('@/lib/validation/validate-user');

const mockError = jest.fn();

describe('SummaryController (e2e)', () => {
  let app: INestApplication;
  let prisma: DeepMockProxy<PrismaClient>;
  let mockUsersService: jest.Mocked<Partial<UsersService>>;
  let mockGoogleAdapter: jest.Mocked<Partial<GoogleAdapter>>;

  beforeAll(() => {
    jest.spyOn(console, 'error').mockImplementation(mockError);
  });

  beforeEach(async () => {
    mockUsersService = {
      findOne: jest.fn(),
    };
    mockGoogleAdapter = {
      verifySite: jest.fn(),
    };

    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule, AuthModule, SummaryModule],
    })
      .overrideProvider(PrismaService)
      .useValue(mockDeep<PrismaClient>())
      .overrideProvider(GoogleAdapter)
      .useValue(mockGoogleAdapter)
      .compile();

    app = moduleFixture.createNestApplication();
    await app.init();
    prisma = moduleFixture.get(PrismaService);
    mockGoogleAdapter = moduleFixture.get(GoogleAdapter);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  afterAll(async () => {
    await app.close();
  });

  describe('/ (GET)', () => {
    describe('positive', () => {
      it('should get a users summary between 2 dates', async () => {
        prisma.diaryDay.findMany.mockResolvedValue(diaryDays_fixture);

        (mockUsersService.findOne as jest.Mock).mockResolvedValue(user_fixture);
        (isValidPassword as jest.Mock).mockResolvedValue(user_fixture);
        (mockGoogleAdapter.verifySite as jest.Mock).mockResolvedValue(true);

        const { access_token } = await mockLoginResponseBody(request, app);

        const response = await request(app.getHttpServer())
          .get('/summary')
          .set('Authorization', 'Bearer ' + access_token)
          .query({
            dateFrom: '2023-04-23',
            dateTo: '2023-04-28',
          });

        expect(response.status).toEqual(200);
        expect(response.body).toMatchSnapshot();
      });

      it('should cater for empty dairy-day data', async () => {
        prisma.diaryDay.findMany.mockResolvedValue([]);

        (mockUsersService.findOne as jest.Mock).mockResolvedValue(user_fixture);
        (isValidPassword as jest.Mock).mockResolvedValue(user_fixture);
        (mockGoogleAdapter.verifySite as jest.Mock).mockResolvedValue(true);

        const { access_token } = await mockLoginResponseBody(request, app);

        const response = await request(app.getHttpServer())
          .get('/summary')
          .set('Authorization', 'Bearer ' + access_token)
          .query({
            dateFrom: '2023-04-23',
            dateTo: '2023-04-28',
          });

        expect(response.status).toEqual(200);
        expect(response.body).toMatchSnapshot();
      });
    });

    describe('negative', () => {
      it('should throw error if no user is found', async () => {
        (mockUsersService.findOne as jest.Mock).mockResolvedValue(null);
        (isValidPassword as jest.Mock).mockResolvedValue(null);
        (mockGoogleAdapter.verifySite as jest.Mock).mockResolvedValue(true);

        const { access_token } = await mockLoginResponseBody(request, app);

        const response = await request(app.getHttpServer())
          .get('/summary')
          .set('Authorization', 'Bearer ' + access_token)
          .query({
            dateFrom: '2023-04-23',
            dateTo: '2023-04-28',
          });

        expect(response.status).toEqual(401);
        expect(response.body).toMatchSnapshot();
      });

      it('should throw error if google verification fails', async () => {
        (mockUsersService.findOne as jest.Mock).mockResolvedValue(user_fixture);
        (isValidPassword as jest.Mock).mockResolvedValue(user_fixture);
        (mockGoogleAdapter.verifySite as jest.Mock).mockResolvedValue(false);

        const { access_token } = await mockLoginResponseBody(request, app);

        const response = await request(app.getHttpServer())
          .get('/summary')
          .set('Authorization', 'Bearer ' + access_token)
          .query({
            dateFrom: '2023-04-23',
            dateTo: '2023-04-28',
          });

        expect(response.status).toEqual(401);
        expect(response.body).toMatchSnapshot();
      });

      it('should throw error if from date is after before date', async () => {
        (mockUsersService.findOne as jest.Mock).mockResolvedValue(user_fixture);
        (isValidPassword as jest.Mock).mockResolvedValue(user_fixture);
        (mockGoogleAdapter.verifySite as jest.Mock).mockResolvedValue(true);

        const { access_token } = await mockLoginResponseBody(request, app);

        const response = await request(app.getHttpServer())
          .get('/summary')
          .set('Authorization', 'Bearer ' + access_token)
          .query({
            dateFrom: '2023-04-28',
            dateTo: '2023-04-23',
          });

        expect(response.status).toEqual(400);
        expect(response.body).toMatchSnapshot();
      });

      it('should throw error if date range is greater than MAX_SUMMARY_MONTH_RANGE = 6', async () => {
        (mockUsersService.findOne as jest.Mock).mockResolvedValue(user_fixture);
        (isValidPassword as jest.Mock).mockResolvedValue(user_fixture);
        (mockGoogleAdapter.verifySite as jest.Mock).mockResolvedValue(true);

        const { access_token } = await mockLoginResponseBody(request, app);

        const response = await request(app.getHttpServer())
          .get('/summary')
          .set('Authorization', 'Bearer ' + access_token)
          .query({
            dateFrom: '2020-04-28',
            dateTo: '2023-04-28',
          });

        expect(response.status).toEqual(400);
        expect(response.body).toMatchSnapshot();
      });

      it('should throw error if dateFrom and dateTo are empty', async () => {
        (mockUsersService.findOne as jest.Mock).mockResolvedValue(user_fixture);
        (isValidPassword as jest.Mock).mockResolvedValue(user_fixture);
        (mockGoogleAdapter.verifySite as jest.Mock).mockResolvedValue(true);

        const { access_token } = await mockLoginResponseBody(request, app);

        const response = await request(app.getHttpServer())
          .get('/summary')
          .set('Authorization', 'Bearer ' + access_token)
          .query({
            dateFrom: '',
            dateTo: '',
          });

        expect(response.status).toEqual(400);
        expect(response.body).toMatchSnapshot();
      });

      it('should throw error if dates are not in date format', async () => {
        (mockUsersService.findOne as jest.Mock).mockResolvedValue(user_fixture);
        (isValidPassword as jest.Mock).mockResolvedValue(user_fixture);
        (mockGoogleAdapter.verifySite as jest.Mock).mockResolvedValue(true);

        const { access_token } = await mockLoginResponseBody(request, app);

        const response = await request(app.getHttpServer())
          .get('/summary')
          .set('Authorization', 'Bearer ' + access_token)
          .query({
            dateFrom: 'bad',
            dateTo: 'input',
            test: 'test',
          });

        expect(response.status).toEqual(400);
        expect(response.body).toMatchSnapshot();
      });
    });
  });
});
