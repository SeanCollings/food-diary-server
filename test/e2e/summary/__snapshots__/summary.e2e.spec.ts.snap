// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SummaryController (e2e) / (GET) negative should throw error if date range is greater than MAX_SUMMARY_MONTH_RANGE = 6 1`] = `
{
  "error": "Bad Request",
  "message": [
    "dateFrom and dateTo max-range is 6 months",
  ],
  "statusCode": 400,
}
`;

exports[`SummaryController (e2e) / (GET) negative should throw error if dateFrom and dateTo are empty 1`] = `
{
  "error": "Bad Request",
  "message": [
    "dateFrom should not be empty",
    "dateFrom must be a valid ISO 8601 date string",
    "dateFrom and dateTo max-range is 6 months",
    "dateFrom must come before dateTo",
    "dateTo should not be empty",
    "dateTo must be a valid ISO 8601 date string",
  ],
  "statusCode": 400,
}
`;

exports[`SummaryController (e2e) / (GET) negative should throw error if dates are not in date format 1`] = `
{
  "error": "Bad Request",
  "message": [
    "dateFrom must be a valid ISO 8601 date string",
    "dateFrom and dateTo max-range is 6 months",
    "dateFrom must come before dateTo",
    "dateTo must be a valid ISO 8601 date string",
  ],
  "statusCode": 400,
}
`;

exports[`SummaryController (e2e) / (GET) negative should throw error if from date is after before date 1`] = `
{
  "error": "Bad Request",
  "message": [
    "dateFrom must come before dateTo",
  ],
  "statusCode": 400,
}
`;

exports[`SummaryController (e2e) / (GET) negative should throw error if google verification fails 1`] = `
{
  "message": "Unauthorized",
  "statusCode": 401,
}
`;

exports[`SummaryController (e2e) / (GET) negative should throw error if no user is found 1`] = `
{
  "message": "Unauthorized",
  "statusCode": 401,
}
`;

exports[`SummaryController (e2e) / (GET) positive should cater for empty dairy-day data 1`] = `
{
  "data": {},
  "dates": [
    "2023-04-23",
    "2023-04-24",
    "2023-04-25",
    "2023-04-26",
    "2023-04-27",
    "2023-04-28",
  ],
  "totalDays": 6,
}
`;

exports[`SummaryController (e2e) / (GET) positive should get a users summary between 2 dates 1`] = `
{
  "data": {
    "2023-04-23": {
      "alcohol": 0,
      "breakfast": [
        "1 cup - breakfast food 2 (description of food)",
      ],
      "dinner": [
        "half bowl - dinner food 1",
      ],
      "lunch": [],
      "snack_1": [],
      "snack_2": [],
      "tea_coffee": 0,
      "water": 2,
    },
    "2023-04-26": {
      "alcohol": 0,
      "breakfast": [
        "1 cup - breakfast food 2 (description of food)",
      ],
      "dinner": [
        "half bowl - dinner food 1",
      ],
      "lunch": [
        "lunch food 1",
        "lunch food 2 (Nice and long destription about this meal that is wonderful to read)",
      ],
      "snack_1": [],
      "snack_2": [],
      "tea_coffee": 0,
      "water": 4,
    },
    "2023-04-28": {
      "alcohol": 1,
      "breakfast": [
        "breakfast food 1",
        "1 cup - breakfast food 2 (description of food)",
      ],
      "dinner": [
        "half bowl - dinner food 1",
        "plate - longer dinner food 2",
        "1/2 - dinner food 3",
        "2 - dinner food 4",
      ],
      "lunch": [
        "lunch food 1",
        "lunch food 2 (Nice and long destription about this meal that is wonderful to read)",
        "4 - lunch food 3",
      ],
      "snack_1": [
        "snack 1 food",
      ],
      "snack_2": [
        "snack 2 food",
      ],
      "tea_coffee": 2,
      "water": 4,
    },
  },
  "dates": [
    "2023-04-23",
    "2023-04-24",
    "2023-04-25",
    "2023-04-26",
    "2023-04-27",
    "2023-04-28",
  ],
  "totalDays": 6,
}
`;
